libraries/c_sdk/standard/mqtt/src/iot_mqtt_serializer_deserializer_wrapper.c

```diff
@@ -0,0 +1,784 @@
+/*
+ * FreeRTOS MQTT V2.1.1
+ * Copyright (C) 2020 Amazon.com, Inc. or its affiliates.  All Rights Reserved.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy of
+ * this software and associated documentation files (the "Software"), to deal in
+ * the Software without restriction, including without limitation the rights to
+ * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
+ * the Software, and to permit persons to whom the Software is furnished to do so,
+ * subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+ * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ *
+ * http://aws.amazon.com/freertos
+ * http://www.FreeRTOS.org
+ */
+
+/**
+ * @file iot_mqtt_serializer_deserializer_wrapper.c
+ * @brief Implements serializer and deserializer wrapper functions for the shim.
+ */
+/* The config header is always included first. */
+#include "iot_config.h"
+
+/* Standard includes. */
+#include <string.h>
+
+/* Error handling include. */
+#include "private/iot_error.h"
+
+/* MQTT internal includes. */
+#include "private/iot_mqtt_internal.h"
+
+/* MQTT v4_beta 2 lightweight library includes. */
+#include "mqtt_lightweight.h"
+
+/* Atomic operations. */
+#include "iot_atomic.h"
+
+/*-----------------------------------------------------------*/
+
+/* Size of Puback packet. */
+#define MQTT_PACKET_PUBACK_SIZE      ( 4 )
+
+/**
+ * @brief Per the MQTT 3.1.1 spec, the largest "Remaining Length" of an MQTT
+ * packet is this value.
+ */
+#define MQTT_MAX_REMAINING_LENGTH    ( 268435455UL )
+
+/*-----------------------------------------------------------*/
+
+/* Generate Id for packet. */
+static uint16_t _nextPacketIdentifier( void );
+
+/*-----------------------------------------------------------*/
+
+/**
+ * @brief Calculate the number of bytes required to encode an MQTT
+ * "Remaining length" field.
+ *
+ * @param[in] length The value of the "Remaining length" to encode.
+ *
+ * @return The size of the encoding of length. This is always `1`, `2`, `3`, or `4`.
+ */
+static size_t _remainingLengthEncodedSize( size_t length );
+
+/*-----------------------------------------------------------*/
+
+/* Generate Id for packet. */
+static uint16_t _nextPacketIdentifier( void )
```

Comment:
This functionality will be removed in Phase 2 approach of Shim Implementation and this file will no longer exists in Phase 2. So that's why kept all the definitions required in this file here itself.